File: src/lib/workspace.js
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';

const WORKSPACE_DIR = path.join(os.homedir(), '.projctl');
const WORKSPACE_FILE = path.join(WORKSPACE_DIR, 'workspace.json');

function load() {
  if (!fs.existsSync(WORKSPACE_FILE)) {
    return { version: 1, projects: [] };
  }
  return JSON.parse(fs.readFileSync(WORKSPACE_FILE, 'utf8'));
}

function save(data) {
  fs.mkdirSync(WORKSPACE_DIR, { recursive: true });
  fs.writeFileSync(WORKSPACE_FILE, JSON.stringify(data, null, 2) + '\n', 'utf8');
}

export function addToWorkspace(projectPath, alias) {
  const resolved = path.resolve(projectPath);
  const data = load();

  const existing = data.projects.find(p => p.path === resolved);
  if (existing) {
    existing.alias = alias || existing.alias;
    existing.lastSeen = new Date().toISOString().slice(0, 10);
  } else {
    data.projects.push({
      path: resolved,
      alias: alias || path.basename(resolved),
      added: new Date().toISOString().slice(0, 10),
      lastSeen: new Date().toISOString().slice(0, 10),
    });
  }

  save(data);
  return data;
}

export function removeFromWorkspace(aliasOrPath) {
  const data = load();
  const resolved = path.resolve(aliasOrPath);
  data.projects = data.projects.filter(
    p => p.alias !== aliasOrPath && p.path !== resolved
  );
  save(data);
  return data;
}

export function listWorkspace() {
  return load().projects;
}

export function getWorkspacePath() {
  return WORKSPACE_FILE;
}

File: projhub/lib/workspace.py
from __future__ import annotations
import json
from datetime import date
from pathlib import Path

_WORKSPACE_DIR = Path.home() / ".projhub"
_WORKSPACE_FILE = _WORKSPACE_DIR / "workspace.json"


def _load() -> dict:
    if not _WORKSPACE_FILE.exists():
        return {"version": 1, "projects": []}
    return json.loads(_WORKSPACE_FILE.read_text(encoding="utf-8"))


def _save(data: dict) -> None:
    _WORKSPACE_DIR.mkdir(parents=True, exist_ok=True)
    _WORKSPACE_FILE.write_text(json.dumps(data, indent=2) + "\n", encoding="utf-8")


def add_to_workspace(project_path: Path, alias: str | None = None) -> dict:
    resolved = Path(project_path).resolve()
    today = date.today().isoformat()
    data = _load()

    existing = next((p for p in data["projects"] if p["path"] == str(resolved)), None)
    if existing:
        if alias:
            existing["alias"] = alias
        existing["lastSeen"] = today
    else:
        data["projects"].append({
            "path": str(resolved),
            "alias": alias or resolved.name,
            "added": today,
            "lastSeen": today,
        })

    _save(data)
    return data


def remove_from_workspace(alias_or_path: str) -> dict:
    resolved = str(Path(alias_or_path).resolve())
    data = _load()
    data["projects"] = [
        p for p in data["projects"]
        if p.get("alias") != alias_or_path and p.get("path") != resolved
    ]
    _save(data)
    return data


def list_workspace() -> list[dict]:
    return _load()["projects"]


def get_workspace_path() -> Path:
    return _WORKSPACE_FILE
